crates/data_components/src/mongodb.rs

```diff
@@ -0,0 +1,286 @@
+use std::any::Any;
+use std::fmt;
+use std::fmt::{Debug, Formatter};
+use std::io::Cursor;
+use std::sync::Arc;
+use arrow::datatypes::SchemaRef;
+use arrow::json::reader::infer_json_schema_from_iterator;
+use arrow::json::ReaderBuilder;
+use async_trait::async_trait;
+use datafusion::catalog::Session;
+use datafusion::common::{project_schema, DataFusionError};
+use datafusion::datasource::{TableProvider, TableType};
+use datafusion::execution::{SendableRecordBatchStream, TaskContext};
+use datafusion::logical_expr::Expr;
+use datafusion::physical_expr::{EquivalenceProperties, Partitioning};
+use datafusion::physical_plan::{DisplayAs, DisplayFormatType, ExecutionMode, ExecutionPlan, PlanProperties};
+use datafusion::physical_plan::stream::RecordBatchReceiverStream;
+use futures::TryStreamExt;
+use mongodb::bson::{to_bson, Bson, Document};
+use mongodb::{Client, Collection};
+use mongodb::options::FindOptions;
+use serde_json::{Value, to_value, from_str};
+use snafu::{ResultExt, Snafu};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to find collection list: {source}"))]
+    CollectionListSearchingError { source: mongodb::error::Error },
```

Comment:
```suggestion
    CollectionListSearching { source: mongodb::error::Error },
```

We don't have this in the style guide, but usually error enum variants omit the `Error` in the variant name.

I'll make a separate update for the style guide to include this note!

Reply:
I see! 
I will apply the style guide.

(cf. I tried to change names of snafu errors in this file on my own before the style guide is updated, but was not satisfied.
for example,`CollectionListSearchingFailure`, `SearchingCollectionListFailed`, which look like just changing postfix `Error` to `Fail`)

Reply:
I have updated the [style guide](https://github.com/spiceai/spiceai/blob/trunk/docs/dev/style_guide.md)!

w.r.t the variant naming, the [error handling guide](https://github.com/spiceai/spiceai/blob/trunk/docs/dev/error_handling.md) might help inform your variant names. The guide talks more about the error messages themselves, but I find it helpful in determining variant name too. For example, `CollectionListSearching` talks about failing to search collection list, so we could name it something like `FailedToListCollection`.

Reply:
I didn't know that there are useful documentations for contributors under `spiceai/docs/dev`!
Thank you for guiding.

I think the variant name which starts with `FailedTo` is intuitive and clean, so I applied those style.